### PR TITLE
Enable extensions

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/5243

@ironcladlou ptal, the first two commits are from https://github.com/openshift/origin/pull/5257
@ncdc This enables the API for the extensions, but doesn't turn on the controllers yet.  Also, I don't remember Solly's(?) nick.

### CODE REVIEW

**Comment by ncdc** [comment]
cc @DirectXMan12

File: pkg/cmd/server/api/v1/types.go [hunk 1]
```diff
@@ -693,6 +693,9 @@ type EtcdConfig struct {
 type KubernetesMasterConfig struct {
 	// APILevels is a list of API levels that should be enabled on startup: v1beta3 and v1 as examples
 	APILevels []string `json:"apiLevels"`
+	// APIGroupVersionBlacklist is a map of groups to the versions (or *) that should be disabled.
+	APIGroupVersionBlacklist map[string][]string `json:"apiGroupVersionBlacklist"`
```
Comment (deads2k) [inline pkg/cmd/server/api/v1/types.go]: @smarterclayton comments about the approach in the config.  Everything on by default.  Easy to disable an entire group.  Easy to disable some versions of a group.

It needs a little plumbing upstream.  Looks like they break if there's more than one version for the group.

File: pkg/cmd/server/api/v1/conversions.go [hunk 2]
```diff
@@ -141,14 +142,49 @@ func init() {
...
+			}
+
+			if out.APIGroupVersionBlacklist == nil {
+				out.APIGroupVersionBlacklist = map[string][]string{}
+			}
+
+			// the APILevels (whitelist) needs to be converted into an internal blacklist
+			if len(in.APILevels) == 0 {
+				out.APIGroupVersionBlacklist[internal.APIGroupKube] = []string{"*"}
+
```
Comment (ironcladlou) [inline pkg/cmd/server/api/v1/conversions.go]: Superfluous newline

File: pkg/cmd/server/api/validation/master.go [hunk 3]
```diff
@@ -432,7 +432,24 @@ func ValidateKubernetesMasterConfig(config *api.KubernetesMasterConfig) Validati
...
+		}
+
+		allowedVersions := sets.NewString(api.KubeAPIGroupsToAllowedVersions[group]...)
+		for i, version := range versions {
+			if version == "*" {
+				continue
+			}
+
+			if !allowedVersions.Has(version) {
+				validationResults.AddWarnings(fielderrors.NewFieldValueNotSupported(fmt.Sprintf(name+"[%d]", i), version, allowedVersions.List()))
```
Comment (ironcladlou) [inline pkg/cmd/server/api/validation/master.go]: Nit: `fmt.Sprintf("%s[%d]", name, i)`

File: pkg/cmd/server/kubernetes/master.go [hunk 4]
```diff
@@ -48,6 +50,10 @@ func (c *MasterConfig) InstallAPI(container *restful.Container) []string {
 		messages = append(messages, fmt.Sprintf("Started Kubernetes API at %%s%s", KubeAPIPrefixV1))
 	}
 
+	if c.Master.EnableExp {
+		messages = append(messages, fmt.Sprintf("Started Kubernetes API Extensions at %%s%s", KubeAPIExtensionsPrefixV1beta1))
```
Comment (ironcladlou) [inline pkg/cmd/server/kubernetes/master.go]: Extra/unintended `%` in format string?

File: pkg/cmd/server/kubernetes/master_config.go [hunk 5]
```diff
@@ -147,14 +149,35 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 		proxyClientCerts = append(proxyClientCerts, clientCert)
 	}
 
+	// TODO you have to know every APIGroup you're enabling or upstream will panic.  It's alternative to panicing is Fataling
+	// It needs a refactor to return errors
 	storageDestinations := master.NewStorageDestinations()
-	storageDestinations.AddAPIGroup("", databaseStorage)
+	// storageVersions is a map from API group to allowed versions that must include all the valid versions
+	// turns out, this must match values from storageDestinations or the master will fatal out, but
+	// the two are directly linked or validated any way upstream.  Upstream seems to be bugged and only allows
```
Comment (ironcladlou) [inline pkg/cmd/server/kubernetes/master_config.go]: Do you mean "the two are *not* directly linked or validated in any way upstream" (emphasis mine)

File: pkg/cmd/server/kubernetes/master_config.go [hunk 6]
```diff
@@ -147,14 +149,35 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
...
+	// a single version per group.
+	// TODO this will fail to compile once we fix upstream
+	storageVersions := map[string]string{}
+
+	enabledKubeVersions := configapi.GetEnabledAPIVersionsForGroup(*options.KubernetesMasterConfig, configapi.APIGroupKube)
+	enabledKubeVersionSet := sets.NewString(enabledKubeVersions...)
+	if len(enabledKubeVersions) > 0 {
+		storageDestinations.AddAPIGroup(configapi.APIGroupKube, databaseStorage)
+		storageVersions[configapi.APIGroupKube] = enabledKubeVersions[0]
```
Comment (ironcladlou) [inline pkg/cmd/server/kubernetes/master_config.go]: Why is it okay to always use the first element?

File: pkg/cmd/server/kubernetes/master.go (hunk 4 above)
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master.go]: > Extra/unintended % in format string?

It gets formatted again.

**Comment by deads2k** [comment]
@ironcladlou  added policy changes in another commit

File: pkg/cmd/server/kubernetes/master_config.go (hunk 6 above)
Comment (deads2k) [inline pkg/cmd/server/kubernetes/master_config.go]: > Why is it okay to always use the first element?

Length is over zero, this arbitrarily chooses the first version until I can fix `storageVersions` upstream to handle multiple versions for one group.

Comment (deads2k) [inline pkg/cmd/server/kubernetes/master_config.go]: > Length is over zero, this arbitrarily chooses the first version until I can fix storageVersions upstream to handle multiple versions for one group.

nm, wrong value here, fixing it.

File: pkg/cmd/server/api/v1/conversions.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/api/v1/conversions.go]: > Unnecessary newline

I like my newline to break apart the else.  Do you feel strongly?

File: pkg/cmd/server/api/validation/master.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/server/api/validation/master.go]: > Nit: fmt.Sprintf("%s[%d]", name, i)

done

File: pkg/cmd/server/api/v1/conversions.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/cmd/server/api/v1/conversions.go]: Not at all

**Comment by deads2k** [comment]
@ironcladlou @ncdc  @DirectXMan12  This works for enabling the API.

[test]

File: pkg/cmd/server/api/helpers.go [hunk 7]
```diff
@@ -498,7 +498,20 @@ func HasOpenShiftAPILevel(config MasterConfig, apiLevel string) bool {
 	return apiLevelSet.Has(apiLevel)
 }
 
-func HasKubernetesAPILevel(config KubernetesMasterConfig, apiLevel string) bool {
-	apiLevelSet := sets.NewString(config.APILevels...)
-	return apiLevelSet.Has(apiLevel)
+func GetEnabledAPIVersionsForGroup(config KubernetesMasterConfig, apiGroup string) []string {
```
Comment (ncdc) [inline pkg/cmd/server/api/helpers.go]: Godoc?

**Comment by deads2k** [comment]
rebased on master

**Comment by smarterclayton** [comment]
Just those comments

File: pkg/cmd/server/api/v1/types.go (hunk 1 above)
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: can we use "disabled" instead of "blacklist"?

**Comment by deads2k** [comment]
[merge]

**Comment by liggitt** [comment]
open ansible issue to update template

**Comment by deads2k** [comment]
> open ansible issue to update template

https://github.com/openshift/openshift-ansible/issues/725

**Comment by deads2k** [comment]
@danmcp how about  making our `needs-rebase` match the kube color?  Those who can distinguish them may find the difference jarring :)

**Comment by danmcp** [comment]
@smarterclayton picked this color

**Pull request merged**